### PR TITLE
GH-38: Add label class for sentences to hold label name and confidence value

### PR DESCRIPTION
**Pull request by tabergma**
related to #38

### CODE REVIEW

**Comment by stefan-it** [comment]
Unfortunately, the `test_text_classifier_trainer.py` fails randomly:

```
stefan@stefan-power-workstation:~/Repositories/github.com/flair/tests$ pytest -k test_text_classifier_trainer.py
======================================================================================================================================================================= test session starts =======================================================================================================================================================================
platform linux -- Python 3.7.0, pytest-3.7.1, py-1.5.4, pluggy-0.7.1
rootdir: /home/stefan/Repositories/github.com/flair, inifile:
collected 51 items / 49 deselected                                                                                                                                                                                                                                                                                                                                

test_text_classifier_trainer.py ..                                                                                                                                                                                                                                                                                                                          [100%]

============================================================================================================================================================ 2 passed, 49 deselected in 13.91 seconds =============================================================================================================================================================
stefan@stefan-power-workstation:~/Repositories/github.com/flair/tests$ 
stefan@stefan-power-workstation:~/Repositories/github.com/flair/tests$ pytest -k test_text_classifier_trainer.py
======================================================================================================================================================================= test session starts =======================================================================================================================================================================
platform linux -- Python 3.7.0, pytest-3.7.1, py-1.5.4, pluggy-0.7.1
rootdir: /home/stefan/Repositories/github.com/flair, inifile:
collected 51 items / 49 deselected                                                                                                                                                                                                                                                                                                                                

test_text_classifier_trainer.py F.                                                                                                                                                                                                                                                                                                                          [100%]

============================================================================================================================================================================ FAILURES =============================================================================================================================================================================
________________________________________________________________________________________________________________________________________________________________ test_text_classifier_single_label ________________________________________________________________________________________________________________________________________________________________

    def test_text_classifier_single_label():
        corpus = NLPTaskDataFetcher.fetch_data(NLPTask.IMDB)
        label_dict = corpus.make_label_dictionary()
    
        glove_embedding: WordEmbeddings = WordEmbeddings('en-glove')
        document_embeddings: DocumentLSTMEmbeddings = DocumentLSTMEmbeddings([glove_embedding], 128, 1, False, 64, False, False)
    
        model = TextClassifier(document_embeddings, label_dict, False)
    
        trainer = TextClassifierTrainer(model, corpus, label_dict, False)
        trainer.train('./results', max_epochs=2)
    
        sentence = Sentence("Berlin is a really nice city.")
    
        for s in model.predict(sentence):
            for l in s.labels:
                assert(l.name is not None)
>               assert(l.confidence > 0.0)
E               assert 0.0 > 0.0
E                +  where 0.0 = NEGATIVE (0.0).confidence

test_text_classifier_trainer.py:28: AssertionError
---------------------------------------------------------------------------------------------------------------------------------------------------------------------- Captured stdout call -----------------------------------------------------------------------------------------------------------------------------------------------------------------------
reading data from %s
----------------------------------------------------------------------------------------------------
epoch 1 - iter 0/1 - loss 0.10860577
----------------------------------------------------------------------------------------------------
TRAIN:  epoch 0 - loss 0.07258360 - f-score 0.2000 - acc 0.2000
DEV:    epoch 0 - loss 0.30907995 - f-score 0.6000 - acc 0.6000
----------------------------------------------------------------------------------------------------
epoch 2 - iter 0/1 - loss 0.08821219
----------------------------------------------------------------------------------------------------
TRAIN:  epoch 1 - loss 0.06576059 - f-score 0.2000 - acc 0.2000
DEV:    epoch 1 - loss 0.27953392 - f-score 0.6000 - acc 0.6000
----------------------------------------------------------------------------------------------------
testing...
MICRO_AVG           	precision: 0.6000 - recall: 0.6000 - accuracy: 0.6000 - f1-score: 0.6000
NEGATIVE            	precision: 0.6000 - recall: 1.0000 - accuracy: 0.6000 - f1-score: 0.7500
POSITIVE            	precision: 0.0000 - recall: 0.0000 - accuracy: 0.6000 - f1-score: 0.0000
----------------------------------------------------------------------------------------------------
======================================================================================================================================================= 1 failed, 1 passed, 49 deselected in 14.19 seconds ========================================================================================================================================================
```

**Comment by alanakbik** [comment]
@stefan-it Hm very strange, it seems to run on my machine with: 

`platform linux -- Python 3.6.4, pytest-3.6.4, py-1.5.4, pluggy-0.7.1
`

**Comment by tabergma** [comment]
@stefan-it I updated the flaky test. Should not happen anymore.

**Comment by stefan-it** [comment]
This is very strange: at least one test fails:

```
stefan@stefan-power-workstation:~/Repositories/github.com/flair/tests$ pytest -k test_text_classifier_trainer.py
======================================================== test session starts ========================================================
platform linux -- Python 3.7.0, pytest-3.7.1, py-1.5.4, pluggy-0.7.1
rootdir: /home/stefan/Repositories/github.com/flair, inifile:
collected 51 items / 49 deselected                                                                                                  

test_text_classifier_trainer.py .F                                                                                            [100%]

============================================================= FAILURES ==============================================================
_________________________________________________ test_text_classifier_mulit_label __________________________________________________

    def test_text_classifier_mulit_label():
        corpus = NLPTaskDataFetcher.fetch_data(NLPTask.IMDB)
        label_dict = corpus.make_label_dictionary()
    
        glove_embedding: WordEmbeddings = WordEmbeddings('en-glove')
        document_embeddings: DocumentMeanEmbeddings = DocumentMeanEmbeddings([glove_embedding], True)
    
        model = TextClassifier(document_embeddings, label_dict, True)
    
        trainer = TextClassifierTrainer(model, corpus, label_dict, False)
        trainer.train('./results', max_epochs=2)
    
        sentence = Sentence("Berlin is a really nice city.")
    
        for s in model.predict(sentence):
            for l in s.labels:
                assert(l.name is not None)
>               assert(0.0 >= l.confidence <= 1.0)
E               assert 0.0 >= tensor(0.7366, grad_fn=<SigmoidBackward>)
E                +  where tensor(0.7366, grad_fn=<SigmoidBackward>) = POSITIVE (0.7365739941596985).confidence

test_text_classifier_trainer.py:51: AssertionError
------------------------------------------------------- Captured stdout call --------------------------------------------------------
reading data from %s
----------------------------------------------------------------------------------------------------
epoch 1 - iter 0/1 - loss 0.14723012
----------------------------------------------------------------------------------------------------
TRAIN:  epoch 0 - loss 0.11114929 - f-score 0.8000 - acc 0.8000
DEV:    epoch 0 - loss 0.17003076 - f-score 0.4000 - acc 0.4000
----------------------------------------------------------------------------------------------------
epoch 2 - iter 0/1 - loss 0.11114930
----------------------------------------------------------------------------------------------------
TRAIN:  epoch 1 - loss 0.10384125 - f-score 0.8000 - acc 0.8000
DEV:    epoch 1 - loss 0.18819386 - f-score 0.4000 - acc 0.4000
----------------------------------------------------------------------------------------------------
testing...
MICRO_AVG           	precision: 0.4000 - recall: 0.4000 - accuracy: 0.4000 - f1-score: 0.4000
NEGATIVE            	precision: 0.0000 - recall: 0.0000 - accuracy: 0.4000 - f1-score: 0.0000
POSITIVE            	precision: 0.4000 - recall: 1.0000 - accuracy: 0.4000 - f1-score: 0.5714
----------------------------------------------------------------------------------------------------
======================================== 1 failed, 1 passed, 49 deselected in 13.58 seconds =========================================
```

and sometimes all test cases:

```
stefan@stefan-power-workstation:~/Repositories/github.com/flair/tests$ pytest -k test_text_classifier_trainer.py
======================================================== test session starts ========================================================
platform linux -- Python 3.7.0, pytest-3.7.1, py-1.5.4, pluggy-0.7.1
rootdir: /home/stefan/Repositories/github.com/flair, inifile:
collected 51 items / 49 deselected                                                                                                  

test_text_classifier_trainer.py FF                                                                                            [100%]

============================================================= FAILURES ==============================================================
_________________________________________________ test_text_classifier_single_label _________________________________________________

    def test_text_classifier_single_label():
        corpus = NLPTaskDataFetcher.fetch_data(NLPTask.IMDB)
        label_dict = corpus.make_label_dictionary()
    
        glove_embedding: WordEmbeddings = WordEmbeddings('en-glove')
        document_embeddings: DocumentLSTMEmbeddings = DocumentLSTMEmbeddings([glove_embedding], 128, 1, False, 64, False, False)
    
        model = TextClassifier(document_embeddings, label_dict, False)
    
        trainer = TextClassifierTrainer(model, corpus, label_dict, False)
        trainer.train('./results', max_epochs=2)
    
        sentence = Sentence("Berlin is a really nice city.")
    
        for s in model.predict(sentence):
            for l in s.labels:
                assert(l.name is not None)
>               assert(0.0 >= l.confidence <= 1.0)
E               assert 0.0 >= tensor(0.1322, grad_fn=<MaxBackward0>)
E                +  where tensor(0.1322, grad_fn=<MaxBackward0>) = POSITIVE (0.13222075998783112).confidence

test_text_classifier_trainer.py:28: AssertionError
------------------------------------------------------- Captured stdout call --------------------------------------------------------
reading data from %s
----------------------------------------------------------------------------------------------------
epoch 1 - iter 0/1 - loss 0.13967187
----------------------------------------------------------------------------------------------------
TRAIN:  epoch 0 - loss 0.09319259 - f-score 0.8000 - acc 0.8000
DEV:    epoch 0 - loss 0.31724125 - f-score 0.4000 - acc 0.4000
----------------------------------------------------------------------------------------------------
epoch 2 - iter 0/1 - loss 0.12188275
----------------------------------------------------------------------------------------------------
TRAIN:  epoch 1 - loss 0.08162130 - f-score 0.8000 - acc 0.8000
DEV:    epoch 1 - loss 0.28208417 - f-score 0.4000 - acc 0.4000
----------------------------------------------------------------------------------------------------
testing...
MICRO_AVG           	precision: 0.4000 - recall: 0.4000 - accuracy: 0.4000 - f1-score: 0.4000
POSITIVE            	precision: 0.4000 - recall: 1.0000 - accuracy: 0.4000 - f1-score: 0.5714
NEGATIVE            	precision: 0.0000 - recall: 0.0000 - accuracy: 0.4000 - f1-score: 0.0000
----------------------------------------------------------------------------------------------------
_________________________________________________ test_text_classifier_mulit_label __________________________________________________

    def test_text_classifier_mulit_label():
        corpus = NLPTaskDataFetcher.fetch_data(NLPTask.IMDB)
        label_dict = corpus.make_label_dictionary()
    
        glove_embedding: WordEmbeddings = WordEmbeddings('en-glove')
        document_embeddings: DocumentMeanEmbeddings = DocumentMeanEmbeddings([glove_embedding], True)
    
        model = TextClassifier(document_embeddings, label_dict, True)
    
        trainer = TextClassifierTrainer(model, corpus, label_dict, False)
        trainer.train('./results', max_epochs=2)
    
        sentence = Sentence("Berlin is a really nice city.")
    
        for s in model.predict(sentence):
            for l in s.labels:
                assert(l.name is not None)
>               assert(0.0 >= l.confidence <= 1.0)
E               assert 0.0 >= tensor(0.7179, grad_fn=<SigmoidBackward>)
E                +  where tensor(0.7179, grad_fn=<SigmoidBackward>) = POSITIVE (0.7179293632507324).confidence

test_text_classifier_trainer.py:51: AssertionError
------------------------------------------------------- Captured stdout call --------------------------------------------------------
reading data from %s
----------------------------------------------------------------------------------------------------
epoch 1 - iter 0/1 - loss 0.13406523
----------------------------------------------------------------------------------------------------
TRAIN:  epoch 0 - loss 0.11122975 - f-score 0.8000 - acc 0.8000
DEV:    epoch 0 - loss 0.16383381 - f-score 0.4000 - acc 0.4000
----------------------------------------------------------------------------------------------------
epoch 2 - iter 0/1 - loss 0.11122975
----------------------------------------------------------------------------------------------------
TRAIN:  epoch 1 - loss 0.10511577 - f-score 0.8000 - acc 0.8000
DEV:    epoch 1 - loss 0.18195871 - f-score 0.4000 - acc 0.4000
----------------------------------------------------------------------------------------------------
testing...
MICRO_AVG           	precision: 0.4000 - recall: 0.4000 - accuracy: 0.4000 - f1-score: 0.4000
POSITIVE            	precision: 0.4000 - recall: 1.0000 - accuracy: 0.4000 - f1-score: 0.5714
NEGATIVE            	precision: 0.0000 - recall: 0.0000 - accuracy: 0.4000 - f1-score: 0.0000
----------------------------------------------------------------------------------------------------
============================================= 2 failed, 49 deselected in 14.36 seconds ==============================================
```
